odoo/modules/registry.py

```diff
@@ -569,6 +569,18 @@ def post_constraint(self, func, *args, **kwargs):
             else:
                 _schema.info(*e.args)
                 self._constraint_queue.append((func, args, kwargs))
+        else:
+            # If a module successfully applies a constraint, we remove all queued constraints with
```

Comment:
You can get there also when the constraint was not applied successully and a new one is has the same arguments.

Reply:
I like what you did in the fixup.

`self._constraint_stacks = defaultdict(dict)` why is this a dict of dicts?
Shouldn't we just have a `self._constraint_stacks = {}` where you replace the value instead of appending it? If we have the same key, we want to apply it only once, don't we?

```
self._constraint_stack.pop(key, None)  # pop the old definition since we have a new one 
try:
   apply_constraint
except ...
  self._constraint_stack[key] = function
```

Reply:
> If we have the same key, we want to apply it only once, don't we?

I wanted to keep old behavior, which I think is correct. If a constraint with same name is enqueued multiple times we try to apply the last one _added_ (that's why I also use move-to-end logic, in case it was repeated). If, starting from the last one, we cannot a apply a constraint we continue and try to apply the previous one. That way we let at least one of the modules apply its constraint, even if a later override cannot be applied.

From the example. Imagine we cannot apply the constraint in `website`, but we can still apply the constraint in `base`. Both failed at their respective load but for some reason once everything was upgraded now the constraint in `base` can be applied, not the one in `website`. If we didn't keep both then we loose the possibility of applying the `base` one.

> self._constraint_stacks = defaultdict(dict) why is this a dict of dicts?

The outer dict is for the keys, the inner one is an OrderedSet with move-to-end logic. AFAICT the standard implementation in Odoo tools doesn't have that method, hence the workaround. Note that the standard implementation also uses a dict ;)

Reply:
For me, it was not completely necessary to re-apply the failed constraints. Since you already have failures, you will have to investigate manually anyways.

Could you add a comment for the first paragraph in the code?
Both solutions are fine for me.

Then, we are good to merge.

Reply:
> For me, it was not completely necessary to re-apply the failed constraints. Since you already have failures, you will have to investigate manually anyways.

I agree, for local development or runbot. OTOH, for upgrades, do we want to let the upgrade get no constraint at all because an override couldn't be added? Unless we cause a hard failure, I'd say we prefer to get one of the constraints even if some override cannot be added.

I added an extra logging line to show all overrides that failed to be added. In this way we have all the information we need to debug the issue right in the end. It also makes it clearer during an upgrade that something didn't finish in a perfect state.

@KangOl could you give some insights here?